pkg/core/resources/apis/meshservice/generate/generator.go

```diff
@@ -0,0 +1,211 @@
+	ctx := user.Ctx(context.Background(), user.ControlPlane)
+
+	for {
+		select {
+		case <-ticker.C:
+			start := time.Now()
+			aggregatedMeshCtxs, err := xds_context.AggregateMeshContexts(ctx, g.resManager, g.meshCache.GetMeshContext)
+			if err != nil {
+				return err
+			}
+			for mesh, meshCtx := range aggregatedMeshCtxs.MeshContextsByName {
+				dataplanes := meshCtx.Resources.Dataplanes()
+				g.generate(ctx, mesh, dataplanes.Items)
+			}
+			g.metric.Observe(float64(time.Since(start).Milliseconds()))
+
+		case <-stop:
+			g.logger.Info("stopping")
+			return nil
+		}
+	}
+}
+
+func (g *Generator) generate(ctx context.Context, mesh string, dataplanes []*core_mesh.DataplaneResource) {
+	meshservicesByName := map[string][]dataplaneAndMeshService{}
+	for _, dataplane := range dataplanes {
+		// TODO order these
+		for name, ms := range g.meshServicesForDataplane(dataplane.Spec) {
+			meshservicesByName[name] = append(meshservicesByName[name], dataplaneAndMeshService{
+				dataplane:   dataplane.GetMeta(),
+				meshService: ms,
+			})
+		}
+	}
+
+	msList := meshservice_api.MeshServiceResourceList{}
+	if err := g.resManager.List(ctx, &msList, store.ListByMesh(mesh)); err != nil {
+		g.logger.Error(err, "failed to list MeshServices")
+	}
+
+	for _, meshService := range msList.Items {
+		conflicting, newMeshService := checkMeshServicesConsistency(meshService.Spec, meshservicesByName[meshService.GetMeta().GetName()])
+		var dps []string
+		for _, dp := range conflicting {
+			dps = append(dps, dp.dataplane.GetName())
+		}
+		if len(conflicting) > 0 {
+			g.logger.Info("conflicting for MeshService", "MeshService", meshService.GetMeta().GetName(), "dps", dps)
+		}
+		delete(meshservicesByName, meshService.GetMeta().GetName())
+		if newMeshService != nil && !reflect.DeepEqual(meshService.Spec, newMeshService) {
+			meshService.Spec = newMeshService
+			if err := g.resManager.Update(ctx, meshService); err != nil {
+				g.logger.Error(err, "couldn't update MeshService", "name", meshService.GetMeta().GetName())
+				continue
+			}
+		}
+	}
+	for name, meshServices := range meshservicesByName {
+		conflicting, newMeshService := checkMeshServicesConsistency(nil, meshServices)
+		meshService := meshservice_api.NewMeshServiceResource()
+		meshService.Spec = newMeshService
+		var dps []string
+		for _, dp := range conflicting {
+			dps = append(dps, dp.dataplane.GetName())
+		}
+		if len(conflicting) > 0 {
+			g.logger.Info("conflicting for MeshService", "MeshService", meshService.GetMeta().GetName(), "dps", dps)
+		}
+		if err := g.resManager.Create(ctx, meshService, store.CreateByKey(name, mesh), store.CreateWithLabels(map[string]string{
+			mesh_proto.ManagedByLabel:      "meshservice-generator",
+			mesh_proto.ResourceOriginLabel: string(mesh_proto.ZoneResourceOrigin),
+		})); err != nil {
+			g.logger.Error(err, "couldn't create MeshService", "name", name)
+			continue
+		}
+	}
+}
+
+func (g *Generator) NeedLeaderElection() bool {
+	return true
+}
```

Comment:
any plans to handle removal? Is it in this PR or other? If other, can we have an issue?

Reply:
you mean of MeshService? fixed

Reply:
that would work for this PR, but I'd improve this with some "delay" mechanism we talked about

Reply:
https://github.com/kumahq/kuma/issues/10975